### PR TITLE
SSE Loss for XGBoost

### PR DESCRIPTION
**Pull request by RishabhGarg108**
This PR attempts to add SSE (Sum of Squared Errors) Loss function for xgboost. The list of methods implemented as of now is:-
1. InitialPrediction - Gives the initial constant prediction used in boosting.
2. Gradients - Calculates the first-order gradients of the loss function. (Both for scalars and vectors.)
3. Hessians - Calculates the second-order gradients of the loss function. (Both for scalars and vectors.)
4. Residuals - Calculates the pseudo residuals of the predictions.

Methods yet to be implemented:-
1. EvaluateGain - Calculates gain for splitting the tree.
2. OutputValue - Calculates the output value for the leaves of the decision tree.

The implementation which I will be doing in this PR will be pretty simple and without any optimizations. Gradually we will keep iterating over things and improving them one at a time.

Also, the methods which are implemented or yet to be implemented (mentioned above) are the basic ones that will be required very frequently. We will keep adding more methods if required later.

CC - @gmanlan @rcurtin

### CODE REVIEW

File: src/mlpack/methods/xgboost/loss_functions/CMakeLists.txt
Comment (rcurtin) [inline src/mlpack/methods/xgboost/loss_functions/CMakeLists.txt:4]: I think you could just use `loss_functions/sse_loss.hpp` in the parent directory's `CMakeLists.txt`, and then you don't need this file. :+1:

File: src/mlpack/methods/xgboost/loss_functions/sse_loss.hpp
Comment (rcurtin) [inline src/mlpack/methods/xgboost/loss_functions/sse_loss.hpp:75]: I *think* (but am not sure) that you can use `VecType::is_row || VecType::is_col` here as the condition. :+1:

File: src/mlpack/methods/xgboost/loss_functions/sse_loss.hpp
Comment (rcurtin) [inline src/mlpack/methods/xgboost/loss_functions/sse_loss.hpp:85]: If that is always true, why not just inline `-Gradients()` wherever you use it?  That seems like it could simplify the code somewhat, and reduce the number of functions that need to be implemented for the loss. :+1:

File: src/mlpack/methods/xgboost/loss_functions/sse_loss.hpp
Comment (rcurtin) [inline src/mlpack/methods/xgboost/loss_functions/sse_loss.hpp:97]: This computes the prediction value for a leaf in a regression tree, right?  Do we use a majority vote for classification, or some other rule?

File: src/mlpack/tests/xgboost_test.cpp
Comment (rcurtin) [inline src/mlpack/tests/xgboost_test.cpp:32]: It wouldn't hurt to add some sanity checks to ensure that, e.g., empty input doesn't crash. :)

File: src/mlpack/methods/xgboost/loss_functions/sse_loss.hpp
Comment (rcurtin) [inline src/mlpack/methods/xgboost/loss_functions/sse_loss.hpp:108]: If I understand the algorithm correctly, I guess this is really the `FitnessFunction` that we will use to evaluate splits.  Maybe we should implement the same API, such that we can use this class as the `FitnessFunction`?  I could see this working if, e.g., we passed a *two*-dimensional matrix (instead of a one-dimensional vector) to the fitness function for the responses/labels, where the two dimensions were the gradients and hessians, respectively.  I believe that this could work since the fitness functions' `Evaluate()` method should be templatized and able to take a matrix instead of a vector.  (If that is not true, it seems like it could be a simple refactoring to make that work.)

**Review by rcurtin — COMMENTED** [review]
Hey @RishabhGarg108, nice work so far!  I left some comments, but, I am not sure of exactly how you are intending to implement things, so, it's possible that some of my comments could be inaccurate or inapplicable.  So, if my ideas don't match at all what you were thinking for implementation, that's ok, mostly I am just writing down ideas and thoughts while I read through the PR. :)

File: src/mlpack/methods/xgboost/loss_functions/CMakeLists.txt
Comment (RishabhGarg108) [reply src/mlpack/methods/xgboost/loss_functions/CMakeLists.txt:4]: Great! I will apply this change.

File: src/mlpack/methods/xgboost/loss_functions/sse_loss.hpp
Comment (RishabhGarg108) [reply src/mlpack/methods/xgboost/loss_functions/sse_loss.hpp:75]: I will give it a shot. If it works, then it will be nice. :)

File: src/mlpack/methods/xgboost/loss_functions/sse_loss.hpp
Comment (RishabhGarg108) [reply src/mlpack/methods/xgboost/loss_functions/sse_loss.hpp:85]: If having so many function calls like this one doesn't have any time cost, then I think that doing it this way will be better. The reason is that from where I see it, all the other operations of a loss function can be expressed in terms of gradients and hessians. So, we can have a base loss function class and all loss functions inherit from it. Then we will only need to implement the gradient and hessian functions for each loss function. 

Maybe we can make the gradient function an inline function? These are just possibilities and I am not exactly sure if it will be the best way to go or not.

File: src/mlpack/methods/xgboost/loss_functions/sse_loss.hpp
Comment (RishabhGarg108) [reply src/mlpack/methods/xgboost/loss_functions/sse_loss.hpp:97]: In the case of XGBoost, regardless of whether we are doing classification or regression, the weak learner is a regression tree. The majority vote kind of thing is done in the gradient boosting machine.

Take the example of binary classification. The weak learner decision tree predicts the probability of a point belonging to class 1. Then the XGBoost class will convert that probability into a class label. Similarly, for multi-class classification, we have to use one-v-one or one-v-rest strategies.

File: src/mlpack/tests/xgboost_test.cpp
Comment (RishabhGarg108) [reply src/mlpack/tests/xgboost_test.cpp:32]: Yeah! Will add the checks 👍

File: src/mlpack/methods/xgboost/loss_functions/sse_loss.hpp
Comment (RishabhGarg108) [reply src/mlpack/methods/xgboost/loss_functions/sse_loss.hpp:108]: Yeah! This could work. The Evaluate() method is already templatized so it should work without any refactoring. I will make the API consistent with the existing FitnessFunction classes.

File: src/mlpack/methods/xgboost/loss_functions/sse_loss.hpp
Comment (rcurtin) [inline src/mlpack/methods/xgboost/loss_functions/sse_loss.hpp:39]: Can you clarify a bit about how or where this will be used?  I'm not sure I understand completely.

File: src/mlpack/methods/xgboost/loss_functions/sse_loss.hpp
Comment (rcurtin) [inline src/mlpack/methods/xgboost/loss_functions/sse_loss.hpp:67]: Do we need an overload for scalars?  The reason that I ask is that it seems like `SSELoss` is implementing a set of methods that could be implemented by a different type of loss too.  That is, `SSELoss` is a template policy meant to be used with a gradient boosting class, and so if a user (or a maintainer) wanted to implement a new type of loss, they would need to implement the same functions that `SSELoss` implements.  That process is much easier when we can reduce the number of functions that need to be implemented to a minimum; so, if a scalar `Hessians()` overload isn't *really* needed and the result could be computed using the regular `Hessians()` overload, I might suggest using that one instead.

But, I have not seen the way that you are intending to use this function, so my comment may be misguided. :)

File: src/mlpack/methods/xgboost/loss_functions/sse_loss.hpp
Comment (rcurtin) [reply src/mlpack/methods/xgboost/loss_functions/sse_loss.hpp:96]: Ok, I guess in this case, I am not really sure where this function will be used.  Is it the objective value that we are minimizing, to be computed after each iteration to see if we have converged?

**Review by rcurtin — COMMENTED** [review]
Hey @RishabhGarg108, looking good!  I left some more comments, hopefully they are helpful.  It is a bit hard to understand SSELoss by itself, because it will really only make sense when integrated with the XGBoost algorithm.  So it is possible my comments don't make that much sense. :smile:

File: src/mlpack/methods/xgboost/loss_functions/sse_loss.hpp
Comment (RishabhGarg108) [reply src/mlpack/methods/xgboost/loss_functions/sse_loss.hpp:39]: Okay, let me go to the basics. In gradient boosting, we start with a constant prediction for all the data points. Then we calculate the residuals (difference between true and predicted values) and then we train multiple trees to fit those residuals.

So to begin with some residuals, we need some initial prediction and what value is given by this function. If you read any paper on gradient boosting, then this initial prediction is calculated by equating the derivative of loss function equal to zero.

I hope I answered your question. :)

File: src/mlpack/methods/xgboost/loss_functions/sse_loss.hpp
Comment (RishabhGarg108) [reply src/mlpack/methods/xgboost/loss_functions/sse_loss.hpp:67]: I completely agree with your point. To be honest, I myself do not have absolute clarity about what functions we will be needing or not. I *think* that we might be needing this overload, so I thought I will directly use it when I will need it. But I might be wrong and we might not need it at all.

There are many things that are specific to loss functions and tasks that they can do. It's a little bit fuzzy area. This might become clearer when we are doing the splitting strategy.

I think maybe we can remove the unnecessary functions later?

**Comment by RishabhGarg108** [comment]
For some reason, I am unable to reply to the last comment.
>Ok, I guess in this case, I am not really sure where this function will be used. Is it the objective value that we are minimizing, to be computed after each iteration to see if we have converged?

This function will be used to calculate the output value for each of the leaf nodes. In the decision tree regressor, it was a straightforward average of the values that end up in that leaf, but in boosting, it depends on our choice of the loss function.

**Comment by RishabhGarg108** [comment]
> Hey @RishabhGarg108, looking good! I left some more comments, hopefully they are helpful. It is a bit hard to understand SSELoss by itself, because it will really only make sense when integrated with the XGBoost algorithm. So it is possible my comments don't make that much sense. 😄

I thought SSE loss is an easy one, that's why I started with it. I think what might be difficult for you is to figure out what and why I have created so many functions. Actually, at this point, my first target is to create a fully working XGBoost class. So, at this moment, I am not worrying about if it is best from the user's perspective or not. I am making so many functions so that I can read through the code easily. For instance, for residuals, it is very easy to read `residuals = - gradients` rather than `residuals = some mathematical function`. We can easily remove some of these functions if we don't need them.

File: src/mlpack/methods/xgboost/loss_functions/sse_loss.hpp
Comment (RishabhGarg108) [reply src/mlpack/methods/xgboost/loss_functions/sse_loss.hpp:107]: @rcurtin I just now realized that we also need regularization parameters lambda and alpha for gain computation in XGBoost. But those parameters are not needed in the normal decision tree. So will it be okay to add these parameters to existing FitnessFunctions for making them consistent?

File: src/mlpack/methods/xgboost/loss_functions/sse_loss.hpp
Comment (rcurtin) [reply src/mlpack/methods/xgboost/loss_functions/sse_loss.hpp:39]: Thanks for the clarification. :+1:

File: src/mlpack/methods/xgboost/loss_functions/sse_loss.hpp
Comment (rcurtin) [reply src/mlpack/methods/xgboost/loss_functions/sse_loss.hpp:67]: Sure, that sounds good, we just need to remember to go back once everything is working and think about whether we chose a good way to do it, or if the abstraction could be improved. :)  (But, you are right, it is very hard to predict the exact right interface to provide, and we may not get it right the first time.  That's okay!)

**Comment by rcurtin** [comment]
Sounds good!  I will stay tuned and wait to see it in use in the XGBoost class. :)

File: src/mlpack/methods/xgboost/loss_functions/sse_loss.hpp
Comment (rcurtin) [inline src/mlpack/methods/xgboost/loss_functions/sse_loss.hpp:93]:
```suggestion
    return -arma::accu(gradients) / (arma::accu(hessians) + lambda);
```
Tiny style fix. :)

**Review by rcurtin — COMMENTED** [review]
To me everything looks good here.  Do you want to merge this now, or is it still a work in progress?

**Comment by RishabhGarg108** [comment]
@rcurtin It is almost complete. Just a couple of things left:-
1. Yesterday, I was searching through the xgboost code to find out how they are doing L1 regularization and I found it. So I will implement that too.
2. Once #3014 is approved, a few minor changes will be required here.

File: src/mlpack/methods/xgboost/loss_functions/sse_loss.hpp
Comment (RishabhGarg108) [inline src/mlpack/methods/xgboost/loss_functions/sse_loss.hpp:129]: Reference - https://github.com/dmlc/xgboost/blob/8e8232fb4cceacb6bb65f61362dd2ed0ad5c3277/src/tree/param.h#L253

The idea here is to shrink the output value closer to using the L1 parameter. That is why there are these conditions.

**Comment by RishabhGarg108** [comment]
Hey @rcurtin @gmanlan, I think this PR is ready now. Earlier, some of the checks were failing due to an error in one of the tests. It is working fine on my system and I am not sure why it was throwing an error. If the error still shows, please take a look and see why is it failing.

The methods implemented in this are:-
1. InitialPrediction
2. OutputLeafValue
3. Evaluate:- To calculate the gain for the node before the split is made. This will be called by `DecisionTreeRegressor::Train` to calculate the `bestGain`. It also initializes the gradients and hessians.
4. Evaluate:- To calculate the gain for the split at an index.

The only thing left here is to write the optimized version of `Evaluate()` similar to the one we did in `MSEGain`. But I am not doing it in this PR. Its implementation will depend largely on how we implement the `NumericSplit` for XGBoost. So, I will do it along with that. I will hopefully open a PR for it today or tomorrow morning (IST).

Although this can be merged right away, but I would recommend that first merge #3014 and then this. The reason is that we might need to change small things here or there according to the changes made there. :)

**Comment by rcurtin** [comment]
Can you try building locally with the CMake configuration flag `-DDEBUG=ON` and see if the tests fail then?  Armadillo size checks aren't done when compiling with debugging off, so you may have a situation where Armadillo fails a size check but the code happens to work anyway if the size check is ignored.

**Comment by RishabhGarg108** [comment]
Yeah! sure. I will compile it with debug option and see if it fails. 👍

**Comment by RishabhGarg108** [comment]
@rcurtin The build is green. I thing we can merge this :)

File: src/mlpack/methods/xgboost/loss_functions/sse_loss.hpp
Comment (rcurtin) [inline src/mlpack/methods/xgboost/loss_functions/sse_loss.hpp:33]:
```suggestion
  SSELoss() : alpha(0), lambda(0) { /* Nothing to do. */}
```
Tiny style fix. :)

**Review by rcurtin — APPROVED** [review]
Looks good to me! Thanks for fixing the build. :+1:

**Pull request merged**
